### PR TITLE
Make Dockerfile a little better

### PR DESCRIPTION
**Pull request by vazir**
Original Dockerfile creating lots of layers by executing multiple RUN and COPY.  Multiple cleanups. Templated username and build dir. Also builds on arm64 ok.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -19,58 +20,44 @@ RUN curl https://models.silero.ai/models/tts/ru/v3_1_ru.pt -o ./silero_model.pt
...
+RUN --mount=target=/var/lib/apt/lists,type=cache,sharing=locked \
+    --mount=target=/var/cache/apt,type=cache,sharing=locked \
+    --mount=type=cache,target=/root/.cache \
+    rm -f /etc/apt/apt.conf.d/docker-clean \
+&& apt update \
+&& apt -y  -o 'Acquire::Retries=3' upgrade \
+&& apt-get install -y -o 'Acquire::Retries=3' \
+    libportaudio2 \
+&& pip install -r ./${FULL_PATH}/requirements-docker.txt \
+&& pip install thefuzz
```
Comment (AlexeyBond) [inline Dockerfile:46]: > Original Dockerfile creating lots of layers by executing multiple RUN and COPY

That did enable efficient docker cache usage, including reuse of layers between different versions.

Current version will re-run all these commands on any change (build wil get much longer) and will make users download new layers containing all of the dependencies with every new version (`docker pull` will take more time and use more internet traffic).

**Comment by vazir** [comment]
No. It will not. Current version create multiple layers and that considered
to be a bad practice. Buildx cashes very effectively and my version also
utilizes apt and pip cacheing.

ср, 22 мар. 2023 г., 09:30 Alexey Bondarenko ***@***.***>:

**Comment by AlexeyBond** [comment]
> multiple layers and that considered to be a bad practice

can you provide any reference for that?

> Buildx cashes very effectively

But still it cannot skip any commands if previous layers have changed.

Btw, [any article](https://snyk.io/blog/best-practices-containerizing-python-docker/#:~:text=2.-,Separate%20dependencies%20from%20source%20code,-This%20second%20best) [on best practices for docker](https://testdriven.io/blog/docker-best-practices/#:~:text=smaller%20attack%20surface.-,Order%20Dockerfile%20Commands%20Appropriately,-Pay%20close%20attention) will include advices like this:

> Always put layers that are likely to change as low as possible in the Dockerfile

**Comment by vazir** [comment]
It's always up to task, but you do not want to make more layers than really needed. Current one becomes just  unreasonably over-layered. If there is a concern for unnecessary rebuild... so while premature optimization is still the root of all evil :) what we just proved, shifting a few lines from the middle of the Dockerfile to the end makes it fully caching the build-stage. so when code is changed we get the 5s for the whole build as the rest is already there :) Will commit the change anyways in to a fork.

```
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 2.41kB done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 transferring context: 80B done
#2 DONE 0.1s

#3 resolve image config for docker.io/docker/dockerfile:1
#3 DONE 0.8s

#4 docker-image://docker.io/docker/dockerfile:1@sha256:39b85bbfa7536a5feceb7372a0817649ecb2724562a38360f4d6a7782a409b14
#4 CACHED

#5 [internal] load metadata for docker.io/curlimages/curl:7.85.0
#5 DONE 0.8s

#6 [internal] load metadata for docker.io/library/python:3.9-slim-bullseye
#6 DONE 0.8s

#7 [stage-1  1/10] FROM docker.io/library/python:3.9-slim-bullseye@sha256:50554b007c60cfc0406278504ce8ab3695b117739905e864ebc4a884931f5c98
#7 DONE 0.0s

#8 [silero-downloader 1/3] FROM docker.io/curlimages/curl:7.85.0@sha256:9fab1b73f45e06df9506d947616062d7e8319009257d3a05d970b0de80a41ec5
#8 DONE 0.0s

#9 [internal] load build context
#9 transferring context: 28.06kB 0.0s done
#9 DONE 0.1s

#10 [stage-1  4/10] RUN --mount=target=/var/lib/apt/lists,type=cache,sharing=locked     --mount=target=/var/cache/apt,type=cache,sharing=locked     --mount=type=cache,target=/root/.cache     rm -f /etc/apt/apt.conf.d/docker-clean && apt update && apt -y  -o 'Acquire::Retries=3' upgrade && apt-get install -y -o 'Acquire::Retries=3'     libportaudio2 && pip install -r .//home/python/irene/requirements-docker.txt && pip install thefuzz
#10 CACHED

#11 [stage-1  5/10] WORKDIR /home/python
#11 CACHED

#12 [silero-downloader 3/3] RUN curl https://models.silero.ai/models/tts/ru/v3_1_ru.pt -o ./silero_model.pt
#12 CACHED

#13 [stage-1  6/10] COPY --link --from=silero-downloader /home/downloader/models/silero_model.pt /home/python/irene/silero_model.pt
#13 CACHED

#14 [stage-1  3/10] COPY --chown=python:python ./requirements-docker.txt /home/python/irene
#14 CACHED

#15 [silero-downloader 2/3] WORKDIR /home/downloader/models
#15 CACHED

#16 [stage-1  2/10] RUN groupadd --gid 1001 python && useradd --create-home python --uid 1001 --gid python && mkdir /home/python/irene
#16 CACHED

#17 [stage-1  7/10] COPY --chown=python:python . /home/python/irene
#17 DONE 0.5s

#18 [stage-1  8/10] RUN cp /home/python/irene/runva_webapi_docker.json /home/python/irene/runva_webapi.json && cp /home/python/irene/docker_plugins/* /home/python/irene/plugins && cp /home/python/irene/plugins_inactive/plugin_tts_silero_v3.py /home/python/irene/plugins
#18 DONE 0.5s

#19 [stage-1  9/10] RUN chown -R python:python /home/python
#19 DONE 0.7s

#20 [stage-1 10/10] WORKDIR /home/python/irene
#20 DONE 0.2s

#21 exporting to image
#21 exporting layers
#21 exporting layers 1.5s done
#21 writing image sha256:9f338efc8203ad608c34e3fca6b34f354ee0b2e0f8fbe5f93e8e2a2bbdb5e7a5 done
#21 naming to docker.io/library/ireneva done
#21 DONE 1.5s

real    0m5,639s
user    0m0,112s
sys     0m0,078s

``` 

> > multiple layers and that considered to be a bad practice
> 
> can you provide any reference for that?
> 
> > Buildx cashes very effectively
> 
> But still it cannot skip any commands if previous layers have changed.
> 
> Btw, [any article](https://snyk.io/blog/best-practices-containerizing-python-docker/#:~:text=2.-,Separate%20dependencies%20from%20source%20code,-This%20second%20best) [on best practices for docker](https://testdriven.io/blog/docker-best-practices/#:~:text=smaller%20attack%20surface.-,Order%20Dockerfile%20Commands%20Appropriately,-Pay%20close%20attention) will include advices like this:
> 
> > Always put layers that are likely to change as low as possible in the Dockerfile

**Comment by AlexeyBond** [comment]
> > > multiple layers and that considered to be a bad practice
> > can you provide any reference for that?
> It was a good practice for years, so brief google search https://www.cloudbees.com/blog/reduce-docker-image-size

just couple of quotes from the article you refer to:

> of course don't try to combine unrelated commands just for the sake of producing a tiny image
> ...
> It’s better to break up layers when adding files (to increase granularity and cacheability)

**Comment by vazir** [comment]
And what you want to say?

> > > > multiple layers and that considered to be a bad practice
> > > > can you provide any reference for that?
> > > > It was a good practice for years, so brief google search https://www.cloudbees.com/blog/reduce-docker-image-size
> 
> just couple of quotes from the article you refer to:
> 
> > of course don't try to combine unrelated commands just for the sake of producing a tiny image
> > ...
> > It’s better to break up layers when adding files (to increase granularity and cacheability)

**Comment by AlexeyBond** [comment]
> And what do you want to say by that? Be specific.

Minimization of number of layers is usually much less prioritized than efficient use of layer cache. And it is confirmed by the article you refer to while explaining opposite order of priorities.

According to my experience, Dockerfiles for relatievely simple packages like this one are created using the following pattern:

1) Install system dependencies, if any - using `apt`/`apk`/other package manager, depending on base image
2) Copy list of dependencies and lock files - `requirements.txt` for Pythin, `package.json`+`package-lock.json` for JS with npm, etc.
3) Install dependencies
4) Copy package sources
5) Build the package (for compiled languages or languages requireing preprocessing, like JS/TS frontend stuff)
6) If significant amount of files produced/copied on previous steps are not necessary in final build then start a new stage and copy necessary files there

Some of sub/sibling packages can be built in separate stages and then copied (with `--link`, so layer containing them may be reused for different versions of final image) into main container as late as possible.

Joining or re-ordering steps 1..3 is not a good idea since lists of dependencies of packages we build usualy are more likely to change than lists of system dependencies needed. Different order of these steps will reduce efficiency of cache.

---

But may be it's all not that much critical for package that is built/downloaded(by each user) once or twice a month. I've spent some time optimizing Dockerfiles for images built maybe up to hundreeds times per day on just few CI servers (that have a lot of other jobs), so I may be overreacting.

**Comment by vazir** [comment]
Looks like you do argue just to argue. Especially for such a simple project. If you get time to argue for the empty, sorry, I do not.

**Comment by AlexeyBond** [comment]
No arguing, just sharing some experience, hoping it will be useful.

**Comment by AlexeyBond** [comment]
Looks like we just have different expectations on how much effort should one put into things they publish in open source. And maybe on reasons why would one do that as well. But that's definitely out of scope of this discussion ¯\_(ツ)_/¯

**Comment by vazir** [comment]
Wow. Your excellency. 

> Looks like we just have different expectations on how much effort should one put into things they publish in open source. And maybe on reasons why would one do that as well. But that's definitely out of scope of this discussion ¯_(ツ)_/¯

**Pull request closed without merging**